### PR TITLE
Ws-609: Upgrade to advanced sentiment

### PR DESCRIPTION
**Pull request by benson-basis**

### CODE REVIEW

File: api/src/test/java/com/basistech/rosette/api/RosetteAPITest.java [hunk 1]
```diff
@@ -403,7 +403,9 @@ public void testGetSentiment() throws IOException {
 
     private void verifySentiment(SentimentResponse response) throws IOException {
         SentimentResponse goldResponse = mapper.readValue(responseStr, SentimentResponse.class);
-        assertEquals(response.getSentiment().size(), goldResponse.getSentiment().size());
+        // this is rather lame.
```
Comment (jkho) [inline api/src/test/java/com/basistech/rosette/api/RosetteAPITest.java]: Remove this type of comment from public rosette-api code

File: model/src/main/java/com/basistech/rosette/apimodel/EntitySentiment.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+/**
+ * Per-entity sentiment info.
+ */
+public class EntitySentiment {
+    private final String mention;
+    private final String type;
+    private final String entityId;
+    private final Label sentiment;
+
+    public EntitySentiment(String mention, String type, String entityId, Label sentiment) {
```
Comment (jkho) [inline model/src/main/java/com/basistech/rosette/apimodel/EntitySentiment.java:38]: Add JavaDoc to all public APIs

File: model/src/main/java/com/basistech/rosette/apimodel/SentimentOptions.java [hunk 3]
```diff
@@ -31,47 +33,28 @@ public SentimentOptions() {
         //
     }
 
-    /**
-     * constructor for {@code SentimentOptions}
-     * @param explain whether to return explanation strings for the sentiment results returned
-     */
-    public SentimentOptions(Boolean explain) {
-        this.explain = explain;
+    public Integer getExplanationCount() {
```
Comment (jkho) [inline model/src/main/java/com/basistech/rosette/apimodel/SentimentOptions.java:38]: Add JavaDoc to APIs

File: model/src/main/java/com/basistech/rosette/apimodel/SentimentResponse.java [hunk 4]
```diff
@@ -16,48 +16,44 @@
...
 
-    private final List<Sentiment> sentiment;
-    
-    /**
-     * constructor for {@code SentimentResponse}
-     * @param sentiment list of sentiment analysis results
-     */
-    public SentimentResponse(List<Sentiment> sentiment) {
-        this.sentiment = sentiment;
+    public SentimentResponse(Label document, Collection<EntitySentiment> entities) {
```
Comment (jkho) [inline model/src/main/java/com/basistech/rosette/apimodel/SentimentResponse.java:34]: Add JavaDoc to APIs

**Comment by jkho** [comment]
Do you want to fix these JavaDoc warnings?

```
[WARNING] Javadoc Warnings
[WARNING] /Users/jkho/git/benson-java/examples/src/main/java/com/basistech/rosette/examples/ExampleBase.java:37: warning: no @return
[WARNING] protected String getApiKeyFromSystemProperty() {
[WARNING] ^
[WARNING] /Users/jkho/git/benson-java/examples/src/main/java/com/basistech/rosette/examples/ExampleBase.java:49: warning: no @return
[WARNING] protected String getAltUrlFromSystemProperty() {
[WARNING] ^
[WARNING] /Users/jkho/git/benson-java/examples/src/main/java/com/basistech/rosette/examples/ExampleBase.java:60: warning: no @param for commandClass
[WARNING] protected static void showUsage(Class<? extends ExampleBase> commandClass) {
[WARNING] ^
[WARNING] /Users/jkho/git/benson-java/examples/src/main/java/com/basistech/rosette/examples/ExampleBase.java:69: warning: no @return
[WARNING] protected static String responseToJson(Response response) throws JsonProcessingException {
[WARNING] ^
[WARNING] /Users/jkho/git/benson-java/examples/src/main/java/com/basistech/rosette/examples/ExampleBase.java:69: warning: no @throws for com.fasterxml.jackson.core.JsonProcessingException
[WARNING] protected static String responseToJson(Response response) throws JsonProcessingException {
[WARNING] ^
```

**Comment by jkho** [comment]
I see lots of updated mock-data (request and reponses) files in the PR. Why is this required for this? Some mock-data files has nothing to do with ASCENT. I must be missing something here.

**Comment by benson-basis** [comment]
There's no way I know of to update a little mock data. You run the python script, and you get what you get. If you have another suggestion, I am all ears (and perhaps a tail).

**Comment by jkho** [comment]
got it..it's been awhile since i touched this.

**Comment by jkho** [comment]
With the rnt & rni end points and response changes. And ASCENT model changes, should we bump the RosetteAPI BINDING_VERSION?

**Comment by benson-basis** [comment]
Before we release 0.10 we should certainly change the version. @guruxu  right now?

**Comment by jkho** [comment]
@lauren12292 can you answer when/which jira should bump BINDING_VERSION before a release?

**Comment by jkho** [comment]
You need to update SentimentOptionsMixin class. It still has the old Boolean explain.

**Comment by benson-basis** [comment]
What I need to do is delete it, since there's no ctor args to mix in for.

File: model/src/main/java/com/basistech/rosette/apimodel/SentimentOptions.java [hunk 5]
```diff
@@ -16,62 +16,51 @@
...
-     * @param explain whether to return explanation strings for the sentiment results returned
-     */
-    public SentimentOptions(Boolean explain) {
-        this.explain = explain;
-    }
-
-    /**
-     * get whether to return explanation strings for the sentiment results returned
-     * @return whether to return explanation strings for the sentiment results returned
+     * @return the maximum number of explanation tokens to return.
```
Comment (jkho) [inline model/src/main/java/com/basistech/rosette/apimodel/SentimentOptions.java:36]: I got confused with the word "tokens". I thought you are limiting the maximum words (tokens) in an explanation. In reality you are limiting the maximum explanations. Suggest to change to

@return the maximum number of explanations to return.

File: model/src/main/java/com/basistech/rosette/apimodel/SentimentOptions.java [hunk 6]
```diff
@@ -16,62 +16,51 @@
...
-        return explain;
+    public Integer getExplanationCount() {
+        return explanationCount;
     }
 
     /**
-     * set whether to return explanation strings for the sentiment results returned
-     * @param explain whether to return explanation strings for the sentiment results returned
+     * Set the maximum number of explanation tokens to return. The default is none.
+     * @param explanationCount the number of tokens.
```
Comment (jkho) [inline model/src/main/java/com/basistech/rosette/apimodel/SentimentOptions.java]: Ditto. number of explanations

File: model/src/main/java/com/basistech/rosette/apimodel/SentimentOptions.java (hunk 5 above)
Comment (benson-basis) [inline model/src/main/java/com/basistech/rosette/apimodel/SentimentOptions.java:36]: I actually thought that an explanation was necessarily a single token, but
I'll change.

**Comment by lauren12292** [comment]
no version number changes until code leaves develop -- to its release tag/branch.

**Comment by jkho** [comment]
@guruxu @benson-basis  I don't have write access to https://github.com/rosette-api/java so I cannot commit the PR. :(

**Pull request merged**